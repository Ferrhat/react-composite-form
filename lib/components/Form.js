import React, {Component} from 'react';
import { isUndefined, fromPairs, defaultTo } from 'lodash';

import {validateField} from '../actions/index';
import FormFieldFactory from '../factory/formFieldFactory';

import { Col, Button, Panel, Row, ButtonToolbar, Glyphicon } from 'react-bootstrap';

class Form extends Component {

    constructor(props) {
        super(props);

        this.state = {
            values: {},
            isVisible: true,
            isValid: true,
        };

        this.valueUpdateHandler = this.valueUpdateHandler.bind(this);
        this.generateFields = this.generateFields.bind(this);
        this.submitHandler = this.submitHandler.bind(this);
        this.toggleVisibility = this.toggleVisibility.bind(this);
        this.resetValues = this.resetValues.bind(this);
        this.validate = this.validate.bind(this);
    }

    componentDidMount() {
        const isValid = this.validate(this.state.values);
        this.setState({ isValid });
    }

    validate(values) {
        const allFieldsAreValid = this.props.fields.every(field => {
            const currentValue = values[field.name];
            const value = isUndefined(currentValue) ? field.value : currentValue;
            const errors = validateField(defaultTo(field.validators, []), value);
            return errors.length == 0;
        });

        return allFieldsAreValid && this.props.customValidator.rule(values);
    }

    valueUpdateHandler(name, value) {
        const values = Object.assign({}, this.state.values, {[name]: value});
        const isValid = this.validate(values);
        this.setState({ values, isValid });
    }

    generateFields() {
        return this.props.fields.map((field, i) => {
            const currentValue = this.state.values[field.name];
            const value = isUndefined(currentValue) ? field.value : currentValue;
            const changeHandler = field.changeHandler ? field.changeHandler : () => {};

            return FormFieldFactory.build(Object.assign({}, field, { valueUpdateHandler: this.valueUpdateHandler, value, changeHandler}), i);
        });
    }

    resetValues() {
        this.setState({
            values: {},
        });
    }

    submitHandler(event) {
        event.preventDefault();
        const defaultValues = fromPairs(this.props.fields.map(field => [field.name, field.value]));
        const values = Object.assign({}, defaultValues, this.state.values);
        this.props.submitHandler(values).then(() => this.resetValues());
    }

    toggleVisibility() {
        this.setState({
            isVisible: !this.state.isVisible,
        });
    }

    render() {
        if (!this.state.isVisible) {
            return <Button onClick={this.toggleVisibility}>{ this.props.title }</Button>;
        }

        if(this.props.version === 'custom') {
            return (
                <div className='container-fluid'>
                    <div className="row">
                        <div className="col">
                            <form className="" onSubmit={this.submitHandler}>
                                <div className="">
                                    {this.generateFields()}
                                    <Button type="submit" bsStyle="primary" disabled={!this.state.isValid}>
                                        {this.props.submitLabel}
                                    </Button>
                                </div>
                            </form>
                        </div>
                    </div>
                </div>
            );
        } else {
            return (
                <Panel className="panel-shadow panel-blue-cap">
                    <Panel.Heading>
                        <Row>
                            <Col xs={8}><span>{this.props.title}</span></Col>
                            <Col xs={4}>
                                <ButtonToolbar className="pull-right" style={{ marginTop: -8 }}>
                                    <Button onClick={this.toggleVisibility}>Hide</Button>
                                </ButtonToolbar>
                            </Col>
                        </Row>
                    </Panel.Heading>
                    <Panel.Body>
                        <form className="form-horizontal" onSubmit={this.submitHandler}>
                            <div className="panel-section">
                                {this.generateFields()}
                                <Button type="submit" bsStyle="primary" className="pull-right" disabled={!this.state.isValid}>
                                    {this.props.submitLabel}
                                </Button>
                            </div>
                        </form>
                    </Panel.Body>
                </Panel>
            );
        }
    }
}

Form.defaultProps = {
    fields: [],
    submitLabel: 'Submit',
    customValidator: { rule: () => true, message: null },
};

export default Form;
