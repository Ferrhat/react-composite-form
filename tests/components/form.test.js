import React from 'react';
import {shallow, mount} from 'enzyme';
import Form from "../../lib/components/Form";


describe('Form', () => {
    const mockSubmitHandler = jest.fn(() => Promise.resolve());
    const mockChangeHandler = jest.fn();
    const form = shallow(<Form submitHandler={mockSubmitHandler} fieldKey={'name'} />);

    it('renders properly', () => {
        expect(form).toMatchSnapshot();
    });

    it('contains a hide and a submit button component', () => {
        expect(form.find('Button').length).toEqual(2);
    });

    it('calls the submithandler function', (done) => {
        const event = {
            preventDefault() {},
            target: { name: 'testName', value: 'testValue' }
        };
        form.setState({values: 'testValue'}, () => {
            expect(form.state('values')).toEqual('testValue');
            form.find('form').simulate('submit', event);
            expect(mockSubmitHandler).toBeCalled();
            process.nextTick(() => {
                expect(form.state('values')).toEqual({});
                done();
            })
        });
    });

    it('generates a text field', () => {
        form.setProps({fields: [{type: 'text', name: 'testField'}]});
        expect(form.find('TextField').length).toEqual(1);
    });

    it('stores the new value of a field', () => {
        const event = {
            preventDefault() {},
            target: { name: 'testName', value: 'testValue' }
        };

        let mounted = mount(<Form submitHandler={mockSubmitHandler} fieldKey={'name'} fields={[{type: 'text', name: 'testField', changeHandler: mockChangeHandler}]} />);
        expect(mounted.state('values')['testField']).toEqual(undefined);
        mounted.find('FormControl').simulate('change', event);
        expect(mounted.state('values')['testField']).toEqual('testValue');
    });

    it('hides the form', () => {
        expect(form.find('Button').length).toEqual(2);
        form.find('Button').first().simulate('click');
        expect(form.find('Button').length).toEqual(1);
    });

    it('validates the form', () => {
        let event = {
            preventDefault() {},
            target: { name: 'testName', value: 'testValue' }
        };

        expect(form.state('isValid')).toEqual(true);
        form.setProps({customValidator: { rule: () => false, message: 'error' }});
        form.simulate('change', event);
        setTimeout(() => {
            expect(form.state('isValid')).toEqual(false);
        }, 1000);
    });
});
