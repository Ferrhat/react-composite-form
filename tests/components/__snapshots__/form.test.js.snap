// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders properly 1`] = `
<Uncontrolled(Panel)
  className="panel-shadow panel-blue-cap"
>
  <PanelHeading
    bsClass="panel"
    componentClass="div"
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        xs={8}
      >
        <span />
      </Col>
      <Col
        bsClass="col"
        componentClass="div"
        xs={4}
      >
        <ButtonToolbar
          bsClass="btn-toolbar"
          className="pull-right"
          style={
            Object {
              "marginTop": -8,
            }
          }
        >
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            disabled={false}
            onClick={[Function]}
          >
            Hide
          </Button>
        </ButtonToolbar>
      </Col>
    </Row>
  </PanelHeading>
  <PanelBody
    bsClass="panel"
    collapsible={false}
  >
    <form
      className="form-horizontal"
      onSubmit={[Function]}
    >
      <div
        className="panel-section"
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="primary"
          className="pull-right"
          disabled={false}
          type="submit"
        >
          Submit
        </Button>
      </div>
    </form>
  </PanelBody>
</Uncontrolled(Panel)>
`;
